### PR TITLE
Deprecate Python global configuration functions [Part 2]

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Continuation of #5923.

Update Python API to allow more flexibility for setting providers and provider options.
1. The providers argument (InferenceSession/TrainingSession constructors, InferenceSession.set_providers()) now also accepts a tuple of  (name, options dict).
2. Fix get_available_providers() API (and the corresponding function in the C API) to return the providers in default priority order. Now it can be used as a starting point for the providers argument and maintain the default priority order.

Convert some usages of the deprecated global configuration functions to use EP-specific options instead.

Update some EP-specific option parsing to fail on unknown options.

Other clean up.

**Motivation and Context**
Deprecate global configuration functions.

### CODE REVIEW

**Comment by edgchen1** [comment]
@KeDengMS, I was planning to update usages of set_nuphar_settings(). However, had some issues with running the current tests. It looks like the Nuphar CI build is broken on master: https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=110
So I'm holding off on updating that until I can run tests to verify the changes.

File: onnxruntime/core/providers/cuda/cuda_execution_provider_info.h [hunk 1]
```diff
@@ -11,12 +11,22 @@
 #include "core/session/onnxruntime_c_api.h"
 
 namespace onnxruntime {
+namespace cuda {
+namespace provider_option_names {
+constexpr const char* kDeviceId = "device_id";
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cuda/cuda_execution_provider_info.h:16]: Can these consts be scoped to the cc file only?

Comment (edgchen1) [reply onnxruntime/core/providers/cuda/cuda_execution_provider_info.h:16]: ok. i thought they could be useful elsewhere, but they aren't used now.

---
In reply to: [546978143](https://github.com/microsoft/onnxruntime/pull/6171#discussion_r546978143) [](ancestors = 546978143)

**Review by pranavsharma — APPROVED** [review]

File: onnxruntime/core/providers/get_execution_providers.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        {
+            kCudaExecutionProvider,
+#ifdef USE_CUDA
+            true,
+#else
+            false,
+#endif
+        },
+        {
+            kMIGraphXExecutionProvider,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/get_execution_providers.cc:37]: MiGraphX and RocM should be above CudaEP  (which is the default for GPU)

**Review by jywu-msft — APPROVED** [review]

**Review by liqunfu — DISMISSED** [review]

**Review by thiagocrepaldi — DISMISSED** [review]

**Review by pranavsharma — APPROVED** [review]

**Review by jywu-msft — DISMISSED** [review]

**Review by thiagocrepaldi — DISMISSED** [review]
Frontend looks good

**Review by KeDengMS — APPROVED** [review]
:shipit:

**Review by liqunfu — APPROVED** [review]

**Pull request merged**